python/lsst/cp/pipe/makeBrighterFatterKernel.py

```diff
@@ -21,131 +21,95 @@
-import lsst.daf.persistence.butlerExceptions as butlerExceptions
-from lsst.cp.pipe.ptc.measurePtcGen2Task import (MeasurePhotonTransferCurveTaskConfig,
-                                                 MeasurePhotonTransferCurveTask)
-from lsst.ip.isr import PhotonTransferCurveDataset
+import lsst.pipe.base.connectionTypes as cT
+
+from lsst.ip.isr import (BrighterFatterKernel)
+from .utils import (funcPolynomial, irlsFit)
+from ._lookupStaticCalibration import lookupStaticCalibration
+
+
+class BrighterFatterKernelSolveConnections(pipeBase.PipelineTaskConnections,
+                                           dimensions=("instrument", "exposure", "detector")):
+    dummy = cT.Input(
+        name="raw",
+        doc="Dummy exposure.",
+        storageClass='Exposure',
+        dimensions=("instrument", "exposure", "detector"),
+        multiple=True,
+        deferLoad=True,
+    )
+    camera = cT.PrerequisiteInput(
+        name="camera",
+        doc="Camera associated with this data.",
+        storageClass="Camera",
+        dimensions=("instrument", ),
+        isCalibration=True,
+        lookupFunction=lookupStaticCalibration,
+    )
+    inputPtc = cT.PrerequisiteInput(
+        name="ptc",
+        doc="Photon transfer curve dataset.",
+        storageClass="PhotonTransferCurveDataset",
+        dimensions=("instrument", "detector"),
+        isCalibration=True,
+    )
 
+    outputBFK = cT.Output(
+        name="brighterFatterKernel",
+        doc="Output measured brighter-fatter kernel.",
+        storageClass="BrighterFatterKernel",
+        dimensions=("instrument", "detector"),
+        isCalibration=True,
+    )
 
-class MakeBrighterFatterKernelTaskConfig(pexConfig.Config):
-    """Config class for bright-fatter effect coefficient calculation."""
 
-    isr = pexConfig.ConfigurableField(
-        target=IsrTask,
-        doc="""Task to perform instrumental signature removal""",
-    )
-    isrMandatorySteps = pexConfig.ListField(
-        dtype=str,
-        doc="isr operations that must be performed for valid results. Raises if any of these are False",
-        default=['doAssembleCcd']
-    )
-    isrForbiddenSteps = pexConfig.ListField(
+class BrighterFatterKernelSolveConfig(pipeBase.PipelineTaskConfig,
+                                      pipelineConnections=BrighterFatterKernelSolveConnections):
+    level = pexConfig.ChoiceField(
+        doc="The level at which to calculate the brighter-fatter kernels",
         dtype=str,
-        doc="isr operations that must NOT be performed for valid results. Raises if any of these are True",
-        default=['doFlat', 'doFringe', 'doBrighterFatter', 'doUseOpticsTransmission',
-                 'doUseFilterTransmission', 'doUseSensorTransmission', 'doUseAtmosphereTransmission']
+        default="AMP",
+        allowed={
+            "AMP": "Every amplifier treated separately",
+            "DETECTOR": "One kernel per detector",
```

Comment:
Does "DETECTOR" mean that you calculate individual kernels per amp, and then take the average of those to produce one kernel for the detector? Or am I misuderstanding?

Reply:
If we are generating a `DETECTOR` level kernel, we run the same per-amplifier loop as in the `AMP` case, but then append all of the scaled cross-correlations into a separate list.  After the loop over amplifiers is done, this set of the cross-correlations is then averaged and successive over-relaxed to produce a final kernel.  This treats each the measurements from each amplifier as another realization of the "true" underlying cross-correlation.  Although this is different than what the calibration method does (there we average the individual final kernels), it seems mathematically better to do it this way when all the data is available.

The prior behavior was to measure the cross-correlation on the full detector, without regard to the amplifier boundaries.  This was likely not optimum either, as there should be zero charge spread between adjacent pixels on the full detector image that come from different amplifiers.  There's also an issue that using proper `AMP` kernels, applied on amplifier boundaries, would probably require masking 8-pixel wide strips along all edges, as the kernel falls off the real pixels.